File: calculator/src/main/java/com/turlir/calculator/Analyzer.java
package com.turlir.calculator;

import com.turlir.calculator.converter.ExpressionExtractor;
import com.turlir.calculator.converter.Member;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

/**
 * Анализирует математическое выражение. Выражение может быть не полным, но должно содержать только известные
 * операторы и {@link Double} операнды, представленные в виде {@link Member}
 */
public class Analyzer {

    private final ExpressionExtractor mConverter;

    public Analyzer(ExpressionExtractor converter) {
        mConverter = converter;
    }

    /**
     * Анализирует математическое выражение
     * @param exp выражение
     * @return последовательность токенов выражения
     * @throws IllegalArgumentException при работе итератора в случае ошибки разбора
     */
    public Iterator<Member> analyze(String exp) {
        return mConverter.iterator(exp);
    }

    /**
     * Анализирует математическое выражение
     * @param exp выражение
     * @return список из токенов выражения
     * @throws IllegalArgumentException в случае ошибки разбора
     */
    public List<Member> slice(String exp) throws IllegalArgumentException {
        Iterator<Member> primaryQ = analyze(exp);
        List<Member> copy = new LinkedList<>();
        while (primaryQ.hasNext()) {
            Member item = primaryQ.next();
            copy.add(item);
        }
        return copy;
    }

}

File: calculator/src/main/kotlin/com/turlir/calculator/Analyzer.kt
package com.turlir.calculator

import com.turlir.calculator.converter.*
import java.util.*

/**
 * Анализирует математическое выражение. Выражение может быть не полным, но должно содержать только известные
 * операторы и [Double] операнды, представленные в виде [Member]
 */
class Analyzer {

    private val mConverter: ExpressionExtractor = MemberConverter()

    /**
     * Анализирует математическое выражение
     * @param exp выражение
     * @return последовательность токенов выражения
     * @throws IllegalArgumentException при работе итератора в случае ошибки разбора
     */
    fun analyze(exp: String): Iterator<Member> = mConverter.iterator(exp)

    /**
     * Анализирует математическое выражение
     * @param exp выражение
     * @return список из токенов выражения
     * @throws IllegalArgumentException в случае ошибки разбора
     */
    @Throws(IllegalArgumentException::class)
    fun slice(exp: String): List<Member> {
        val primaryQ = analyze(exp)
        val copy = LinkedList<Member>()
        while (primaryQ.hasNext()) {
            val item = primaryQ.next()
            copy.add(item)
        }
        return copy
    }

    /**
     * Анализирует математическое выражение
     * @param exp выражение
     * @return связный список из токенов математического выражения или {@link MathExpression} если выражение пустое
     * @throws IllegalArgumentException в случае ошибки разбора
     */
    @Throws(IllegalArgumentException::class)
    fun expression(exp: String): MathExpression {
        val primaryQ = analyze(exp)
        if(!primaryQ.hasNext()) throw EmptyExpressionException()
        return Expression(primaryQ)
    }

    /**
     * Анализирует математическое выражение
     * @param exp выражение
     * @return связный список из токенов математического выражения или {@link MathExpression} если выражение пустое.
     * В отличие от метода expression(String) возвращает контейнер с лениво вычисляемыми токенами
     * @throws IllegalArgumentException в случае ошибки разбора
     */
    @Throws(IllegalArgumentException::class)
    fun sequenceExpression(exp: String): MathExpression {
        val primaryQ = analyze(exp)
        if(!primaryQ.hasNext()) throw EmptyExpressionException()
        return SequenceExpression(primaryQ)
    }

}
